src/ICS20Transfer.sol

```diff
@@ -72,59 +67,49 @@ contract ICS20Transfer is IIBCApp, IICS20Transfer, IICS20Errors, Ownable, Reentr
             revert ICS20MsgSenderIsNotPacketSender(msg_.sender, packetData.sender);
         }
 
-        _transferFrom(packetData.sender, address(this), packetData.erc20ContractAddress, packetData.amount);
+        // transfer the tokens to us (requires the allowance to be set)
+        _transferFrom(packetData.sender, address(this), packetData.erc20Contract, packetData.amount);
+
+        // if the denom is prefixed by the port and channel on which we are sending
+        // the token, then we must be returning the token back to the chain they originated from
+        if (packetData.receiverChainIsSource) {
+            // receiver chain is source: burn the vouchers
+            // TODO: Implement escrow balance tracking (#6)
+            IBCERC20 ibcERC20Contract = IBCERC20(packetData.erc20Contract);
+            ibcERC20Contract.burn(packetData.amount);
+        }
 
         emit ICS20Transfer(packetData);
     }
 
     function onRecvPacket(OnRecvPacketCallback calldata msg_) external onlyOwner nonReentrant returns (bytes memory) {
-        // TODO Emit error event
+        // Since this function mostly returns acks, also when it fails, the ics26router (the caller) will log the ack
         if (keccak256(abi.encodePacked(msg_.packet.version)) != keccak256(abi.encodePacked(ICS20Lib.ICS20_VERSION))) {
             return ICS20Lib.errorAck(abi.encodePacked("unexpected version: ", msg_.packet.version));
```

Comment:
Since this is the first check and it returns I was thinking it was OK to just return the ack. I have tried to follow the logic of the ibc-go implementation for acks, where they return errors and in the top level handler an error is just turned into an error ack:
```go
if ackErr = im.keeper.OnRecvPacket(ctx, packet, data); ackErr != nil {
	ack = channeltypes.NewErrorAcknowledgement(ackErr)
	im.keeper.Logger(ctx).Error(fmt.Sprintf("%s sequence %d", ackErr.Error(), packet.Sequence))
	return ack
}
```

But happy to get some more input from Aditya as well :)

Reply:
So a revert would revert the entire tx correct?

In this case, that would leave a second relayer open to submitting the exact same packet again and reaching the same revert.

This is why we return error acks on application error. That way we have a record of processing the packet and the error ack can be processed by the other side. It prevents infinite replays of the same error and also we want to give the sender chain an error ack saying your version is invalid rather than just having it timeout correct?

Reply:
Are there any scenarios where we would want to revert the entire tx? If not, would it be better revert in ICS20Transfer (instead of returning the ack) and catch any errors in the router? It is a bit ugly, but some reverts might be hard to avoid without adding custom error returns and handling everywhere else.